Code review recommended practices (#727)
microsoft/code-with-engineering-playbook: docs/code-reviews/process-guidance/author-guidance.md

- If you don't understand a comment, ask questions in the review itself as opposed to a private chat
- If a thread gets bloated without a conclusion, have a meeting with the reviewer (call them or knock on door)

## Use checklists

> At the moment of creating a pull request it is a very good idea to put in the description of it a checklist of expected objectives to achieve in the PR. This way reviewers can be able to focus on specific areas and also consider some other tasks as secondary according to this list.

Reviewer: ```suggestion
At the moment of creating a pull request it is a good idea to put in the description of it a checklist of expected objectives to achieve in the PR. This way reviewers can be able to focus on specific areas and also consider some other tasks as secondary according to this list.
```

## Code should have annotations before the review

> If big files are something that can not be avoided then you need to include comments in order to make it easier to the teams to review, with clear comments developers can go faster in the process and also can identify goals for every code block.

Reviewer: ```suggestion
If you can't avoid large PRs, include explanations of the changes in order to make it easier for the reviewer to review the code, with clear comments the reviewer can identify the goal of every code block.
```

Reviewer: Must say that I don't necessarily agree with this. The code should be able to speak for itself, through proper naming and through splitting large functions into smaller etc. If we need to add comments, the code should be refactored to be more readable. Comments will always get stale and incorrect over time as the code changes, but the devs forget about updating the comments.

We should however, add comments to the PR to describe the changes we have made, and why.

Author: I like the improvement, change made!

## Track progress

- If the reviewers have not responded in a reasonable time (generally a day or two), ping them or raise the issue in a daily meeting.

## Foster a positive code review culture

> This is the easiest recommendation to follow and yet the most absent form code reviews. Code reviews play a critical role in product quality and it should not represent an arena for long discussions or even worse a battle of egos. What matters is a bug caught not who made it, not who found it, not who solved it. The only thing that matters is having the best possible product.

Reviewer: ```suggestion
Code reviews play a critical role in product quality and it should not represent an arena for long discussions or even worse a battle of egos. What matters is a bug caught not who made it, not who found it, not who solved it. The only thing that matters is having the best possible product.
```

### Get a process for fixing found issues

Before getting started with any commit, your team should define how you will measure the effectiveness of peer review and get some goals. Using a [SMART criteria](https://en.wikipedia.org/wiki/SMART_criteria) the team can get this objectives in the most accurate way and not something very general like **fix more bugs**, the more precise a goal can be the easier it will be to hit the target.

Metrics can play a significant role in this part, you can include an **Inspection rate** (the speed with which review is performed), **Defect rate** (the number of bugs found within an hour of review) and **Defect density** (the number of bugs found per piece of code) as the most common ones to be considered.

## Code should have annotations before the review

If you can't avoid large PRs, include explanations of the changes in order to make it easier for the reviewer to review the code, with clear comments the reviewer can identify the goal of every code block.

## Track progress

- If the reviewers have not responded in a reasonable time (generally a day or two), ping them or raise the issue in a daily meeting.

> ## Foster a positive code review culture

Reviewer: This is more a reviewer guidance than an author guidance - was it put here by mistake?

Author: Well, goal here was to let the author know it's ok to ping the reviewer and let him know there's some code there to be reviewed and probably unblock this in a faster way.

Reviewer: Are we referring to the same section here? This was in regards to 

"Foster a positive code review culture

Code reviews play a critical role in product quality and it should not represent an arena for long discussions or even worse a battle of egos. What matters is a bug caught not who made it, not who found it, not who solved it. The only thing that matters is having the best possible product."

Author: Well, not that you say it that way yes, it seems to fit.

- If you don't understand a comment, ask questions in the review itself as opposed to a private chat
- If a thread gets bloated without a conclusion, have a meeting with the reviewer (call them or knock on door)

## Use checklists

At the moment of creating a pull request it is a good idea to put in the description of it a checklist of expected objectives to achieve in the PR. This way reviewers can be able to focus on specific areas and also consider some other tasks as secondary according to this list.

### Get a process for fixing found issues

> Before getting started with any commit, your team should define how you will measure the effectiveness of peer review and get some goals. Using a [SMART criteria](https://en.wikipedia.org/wiki/SMART_criteria) the team can get this objectives in the most accurate way and not something very general like **fix more bugs**, the more precise a goal can be the easier it will be to hit the target.

Reviewer: This should probably go in the general section

Metrics can play a significant role in this part, you can include an **Inspection rate** (the speed with which review is performed), **Defect rate** (the number of bugs found within an hour of review) and **Defect density** (the number of bugs found per piece of code) as the most common ones to be considered.

### Code should have annotations before the review

If you can't avoid large PRs, include explanations of the changes in order to make it easier for the reviewer to review the code, with clear comments the reviewer can identify the goal of every code block.

> ### Make comments clear

Reviewer: This is reviewer guidance - not author guidance

Author: After talking to some of our teammates we think it fits in author, we just gave it more context.

- Explain why the code needs to change.
- Prefix a “point of polish” with “Nit:”.
- Suggest changes to a PR by using the suggestion feature (available in [GitHub](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request) and Azure DevOps) or by creating a PR to the author branch.
- If one or two comments don’t resolve a disagreement, talk in person or create group discussion.

> ### Foster a positive code review culture

Reviewer: This should probably go in the reviewer section - or in the general section
